package firmament;

import "base/reference_desc.proto";
import "base/resource_vector.proto";

message TaskDescriptor {
  enum TaskState {
    CREATED = 0;
    BLOCKING = 1;
    RUNNABLE = 2;
    ASSIGNED = 3;
    RUNNING = 4;
    COMPLETED = 5;
    FAILED = 6;
    ABORTED = 7;
    DELEGATED = 8;
    UNKNOWN = 9;
  }

  enum TaskType {
    SHEEP = 0;
    RABBIT = 1;
    DEVIL = 2;
    TURTLE = 3;
  }

  required uint64 uid = 1;
  required string name = 2;
  required TaskState state = 3;
  optional string job_id = 4;
  optional uint64 index = 5;
  /* Inputs/outputs */
  repeated ReferenceDescriptor dependencies = 6;
  repeated ReferenceDescriptor outputs = 7;
  /* Command and arguments */
  required bytes binary = 8;
  repeated string args = 9;
  /* Children */
  repeated TaskDescriptor spawned = 10;
  /* Runtime meta-data */
  optional string last_heartbeat_location = 11;
  optional uint64 last_heartbeat_time = 12;
  /* Delegation info */
  optional string delegated_to = 13;
  optional string delegated_from = 14;
  /* Timestamps */
  optional uint64 submit_time = 15;
  optional uint64 start_time = 16;
  optional uint64 finish_time = 17;
  /* Deadlines */
  optional uint64 relative_deadline = 18;
  optional uint64 absolute_deadline = 19;
  optional uint64 equivalence_class = 20;
  /* Application-specific fields */
  /* TODO(malte): move these to sub-messages */
  optional uint64 port = 21;
  optional uint64 input_size = 22;
  /* TaskLib related stuff */
  optional bool inject_task_lib = 23 [default = false];
  /* Task resource request and priority */
  optional ResourceVector resource_request = 24;
  optional uint32 priority = 25;
  optional TaskType task_type = 26;
}
